### ISSUE TITLE
Patchwork stuck [sbot locking]

### ISSUE THREAD

**Issue by pferreir** [ssbc/patchwork#597]
I'm not sure whether this is a Patchwork or scuttlebot issue, but decided to file it here. Basically, every now and then Patchwork will get stuck (e.g. when trying to open a conversation thread) and only after a couple of minutes it will load the content. And the "Downloading new messages" indicator at the top somehow doesn't go away. If i restart the client everything is fine, though.

There is at least [one other person](https://viewer.scuttlebot.io/%25wv%2FIenEk85cU96A7p%2F5%2BC5lj5qhOCdzTdPteyOSilB4%3D.sha256) experiencing this, so I guess it's not misconfiguration on my side. I tried wiping my SSB DB and starting from scratch (in case something has been corrupted) and reinstalling patchwork, same problem.

**Comment by mmckegg** [comment]
Yes this is a known issue with sbot (only when using legacy replication, not ebt). It is the "dreaded locking" referred to in #589

See mega thread on ssb where I try to debug it: [%GbJVNVx4jcO6kmrdrRsZ3RUyXeoPRKKf8vsM6s5v0EQ=.sha256](https://viewer.scuttlebot.io/%25GbJVNVx4jcO6kmrdrRsZ3RUyXeoPRKKf8vsM6s5v0EQ%3D.sha256)

This is triggered when a remote peer closes the connection without warning, and then your side tries to cleanup the closed connection. It can also be triggered by turning off your wifi.

This bug has existed for at least a year. And affects ALL CLIENTS. But it has only become noticeable now that there are so many FOAFs (friend-of-friends). Also, patchwork used to buffer a long way ahead in the feed (1000) which disguised this, now it only reads 50 messages ahead.

I thought I'd fixed it in the last release, so it momentarily went away, but turns out I just added a massive memory leak instead.

I've spent 12 hours+ trying to fix this damn bug. It is impossible to get to the bottom of and I am completely sick of it. Only @dominictarr can fix this (maybe).

The easy answer is we just switch to use ebt everywhere.

**Comment by mmckegg** [comment]
Strange, this bug is no longer reproducible when I disconnect from wifi.

Very mysterious! Maybe a recent OS update fixed it?

What OS and version are you running @pferreir?

**Comment by pferreir** [comment]
I'm running Fedora 25.

**Comment by Marak** [comment]
I think I've seen something like this happen occasionally on MacOS.

I've seen it happen when I close the machine and re-open it. Sometimes Patchwork will just spin forever and not update messages. Restarting the app seems to fix the problem.

I'll keep an eye out and see if I can reproduce.

**Comment by mixmix** [comment]
## capturing history from matt
 
- has spent 10 hours on this to date
- sbot bug : 
  - can get sbot to lock when run by itself
  - ... somewhere deep, maybe muxrpc?
- has been present for ages, particularly obvious in 3.6.6
- went away in 3.6.4 ..? did something which kinda fixed this bug

## symptoms

these are probably related 

- feed stops loading (scrolling stops for a minute)
- you click on message and nothing happens immediately, maybe after a minute +
- progress bars freeze (sidebar with pubs, replication / indexing progress)
- you try to use an invite and nothing happens (problem happening on pub)
- you try to run `sbot whoami` in command line... it can't connect.
  - you can run this in a loop to see when it happens

## reproducing the bug

- remote peer suddenly disconnecting (maybe?)
- turning off wifi (used to work, doesn't any more)
- worse with more people (recently)

**Comment by dcposch** [comment]
i may have run into this while trying to use patchwork earlier today.

*version: patchwork 3.6.6*

multiple times, it just got stuck in an infinite loading state:

*steps: open the app (rare bug, usually works)*

![screen shot 2017-08-13 at 2 52 57 pm](https://user-images.githubusercontent.com/169280/29258487-f36cba3a-806e-11e7-9e3a-fa3fc03caeba.png)

*steps: open the app, click on Profile (rare bug, usually works)*

![screen shot 2017-08-13 at 2 52 23 pm](https://user-images.githubusercontent.com/169280/29258495-02ba4a52-806f-11e7-9bd9-9e45475f5425.png)

each time, the problem resolved after i quit and re-opened Patchwork

i opened DevTools to try to find a stacktrace, but there was no output in the console and nothing in the network tab. see https://github.com/ssbc/patchwork/issues/603

cc @dominictarr

**Comment by mixmix** [comment]
Thanks for the clear documentation. I think we need to mark 3.6.6 unstable.

Please try 3.6.4

**Comment by dominictarr** [comment]
@dcposch thanks for the detailed report. I was able to detect and fix a problem like this before, and added the progress bar stuff (which where used to monitor it's occurance) how often would you say it's happening?

**Comment by mmckegg** [comment]
Just made a very interesting discovery in regards to this!

Downgrading to Electron **1.3** seems to fix the locking! This version of electron uses node **6.5** (which I noticed when running directly does not experience the locking either). These older versions do lock for a brief moment (maybe up to 1 second), but nothing like the multiple minutes seen in electron 1.4 and above!

I'm just about to test out Node 8.0 and see if it does the locking when standalone.

**Comment by mmckegg** [comment]
Latest version of Node works fine. Looks like it was just the 7.x line that had problems (which includes all versions of electron 1.4 through 1.7).

---

But some very interesting news: **Electron 1.8.1** has just been released (based on Node 8.2.1) and it seems to have fixed the locking issue! At least my wifi disable hack no longer triggers it. Will run for a few days and see if it is cured! 🤞 heres hoping 🙏 

Unfortunately Electron 1.8 is suffering from a lack of prebuilds (**abi version 57**). There is no sodium-native prebuild for it. /cc @emilbayes

**Comment by mixmix** [comment]
Awesome. Are you doing this on a branch Matt? Would be great for a bunch of
us to test it.
I can test win and Linux

**Comment by mmckegg** [comment]
@mixmix 

It's on master now!

The only catch is that you currently have to manually copy over the prebuilds for `sodium-native` for electron 1.8.

```sh
git pull origin master

# burn them all
rm -rf node_modules
rm package-lock.json
npm install

# copy across node prebuild
cd node_modules/sodium-native/prebuilds/YOUR_PLATFORM
cp node-57.node electron-57.node
```

**Comment by mmckegg** [comment]
I think this issue is now mostly resolved in 3.7.1

Going to close this issue until someone complains!
